sklearn/linear_model/_coordinate_descent.py

```diff
@@ -146,47 +149,35 @@ def _alpha_grid(
             "your estimator with the appropriate `alphas=` "
             "argument."
         )
-    n_samples = len(y)
-
-    sparse_center = False
-    if Xy is None:
-        X_sparse = sparse.isspmatrix(X)
-        sparse_center = X_sparse and (fit_intercept or normalize)
-        X = check_array(
-            X, accept_sparse="csc", copy=(copy_X and fit_intercept and not X_sparse)
+    if Xy is not None and ((sample_weight is None) or (fit_intercept is False)):
+        Xyw = Xy
+    else:
+        X, y, X_offset, _, _ = _preprocess_data(
+            X,
+            y,
+            fit_intercept,
+            normalize=normalize,
+            copy=copy_X,
+            sample_weight=sample_weight,
+            check_input=False,
         )
-        if not X_sparse:
-            # X can be touched inplace thanks to the above line
-            X, y, _, _, _ = _preprocess_data(X, y, fit_intercept, normalize, copy=False)
-        Xy = safe_sparse_dot(X.T, y, dense_output=True)
-
-        if sparse_center:
-            # Workaround to find alpha_max for sparse matrices.
-            # since we should not destroy the sparsity of such matrices.
-            _, _, X_offset, _, X_scale = _preprocess_data(
-                X, y, fit_intercept, normalize
-            )
-            mean_dot = X_offset * np.sum(y)
-
-    if Xy.ndim == 1:
-        Xy = Xy[:, np.newaxis]
-
-    if sparse_center:
-        if fit_intercept:
-            Xy -= mean_dot[:, np.newaxis]
-        if normalize:
-            Xy /= X_scale[:, np.newaxis]
+        if sample_weight is not None:
+            yw = y * sample_weight / sample_weight.mean()
+        else:
+            yw = y
+        if sparse.issparse(X):
+            Xyw = safe_sparse_dot(X.T, yw, dense_output=True) - np.sum(yw) * X_offset
+        else:
+            Xyw = np.dot(X.T, yw)
 
-    alpha_max = np.sqrt(np.sum(Xy**2, axis=1)).max() / (n_samples * l1_ratio)
+    if Xyw.ndim == 1:
+        Xyw = Xyw[:, np.newaxis]
+    n_samples = X.shape[0]
```

Comment:
Are you sure it is `n_samples` here, and not `sample_weight.sum()`?

Reply:
Above in the code, we define `yw = y * sample_weight / sample_weight.mean()`. This allows us to define `n_samples` in a way that works regardless of whether sample weights are used.

(More generally, I'm willing to bet sklearn will eventually make the package-wide choice to normalize all sample weights to have mean 1. There seems to be some foot dragging on this matter for the sake of "backward-compatibility", despite the fact that almost every major release breaks backward-compatibility in some way or another....)

Reply:
Removed the previous `sample_weight / sample_weight.mean()` so that we can use `sample_weight.sum()` here.